app/controllers/api/storage.php

```diff
@@ -555,135 +408,271 @@
+        // Check if file size is exceeding allowed limit
+        $fileSizeValidator = new FileSize($maximumFileSize);
+        if (!$fileSizeValidator->isValid($fileSize)) {
             throw new Exception('File size not allowed', 400);
         }
-
-        $device = Storage::getDevice('files');
-
-        if (!$upload->isValid($file['tmp_name'])) {
+        
+        $upload = new Upload();
+        if (!$upload->isValid($fileTmpName)) {
             throw new Exception('Invalid file', 403);
         }
 
         // Save to storage
-        $size = $device->getFileSize($file['tmp_name']);
-        $path = $device->getPath(\uniqid().'.'.\pathinfo($file['name'], PATHINFO_EXTENSION));
-        $path = $bucket->getId() . '/' . $path;
-        
-        if (!$device->upload($file['tmp_name'], $path)) { // TODO deprecate 'upload' and replace with 'move'
-            throw new Exception('Failed moving file', 500);
+        $fileSize ??= $deviceLocal->getFileSize($fileTmpName);
+        $path = $deviceFiles->getPath($fileId . '.' . \pathinfo($fileName, PATHINFO_EXTENSION));
+        $path = str_ireplace($deviceFiles->getRoot(), $deviceFiles->getRoot() . DIRECTORY_SEPARATOR . $bucket->getId(), $path); // Add bucket id to path after root
+
+        if($permissionBucket) {
+            $file = Authorization::skip(function() use ($dbForProject, $bucketId, $fileId) {
+                return $dbForProject->getDocument('bucket_' . $bucketId, $fileId);
+            });
+        } else {
+            $file = $dbForProject->getDocument('bucket_' . $bucketId, $fileId);
         }
 
-        $mimeType = $device->getFileMimeType($path); // Get mime-type before compression and encryption
+        $metadata = ['content_type' => $deviceLocal->getFileMimeType($fileTmpName)];
+        if (!$file->isEmpty()) {
+            $chunks = $file->getAttribute('chunksTotal', 1);
+            $metadata = $file->getAttribute('metadata', []);
+            if ($chunk === -1) {
+                $chunk = $chunks;
+            }
+        }
+
+        $chunksUploaded = $deviceFiles->upload($fileTmpName, $path, $chunk, $chunks, $metadata);
+        if (empty($chunksUploaded)) {
+            throw new Exception('Failed uploading file', 500);
+        }
 
-        if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $size <= APP_LIMIT_ANTIVIRUS) {
-            $antivirus = new Network(App::getEnv('_APP_STORAGE_ANTIVIRUS_HOST', 'clamav'),
+        $read = (is_null($read) && !$user->isEmpty()) ? ['user:' . $user->getId()] : $read ?? [];
+        $write = (is_null($write) && !$user->isEmpty()) ? ['user:' . $user->getId()] : $write ?? [];
+        if ($chunksUploaded === $chunks) {
+            if (App::getEnv('_APP_STORAGE_ANTIVIRUS') === 'enabled' && $bucket->getAttribute('antiVirus', true) && $fileSize <= APP_LIMIT_ANTIVIRUS && App::getEnv('_APP_STORAGE_DEVICE', Storage::DEVICE_LOCAL) === Storage::DEVICE_LOCAL) {
```

Comment:
We switched from `AntiVirus` spelling to `Antivirus`. So here it should be `antivirus` for attribute, I think. WDYT? Can we make sure the spelling is correct throughout the whole PR?

```
let promise = health.getAntivirus();
```

Reply:
done